zilliqa/src/api/eth.rs

```diff
@@ -162,110 +174,117 @@ fn get_storage_at(params: Params, node: &Arc<Mutex<Node>>) -> Result<String> {
-            .unwrap()
-            .get_account(Address(address), block_number)?
-            .nonce
-            .to_hex()
-    );
-
-    Ok(node
+    let nonce = node
         .lock()
-        .unwrap()
-        .get_account(Address(address), block_number)?
+        .await
+        .get_account(Address(address), block_number)
+        .await?
         .nonce
-        .to_hex())
+        .to_hex();
+    trace!("get_transaction_count resp: {:?}", nonce);
+
+    Ok(nonce)
 }
 
-fn get_gas_price(_: Params, node: &Arc<Mutex<Node>>) -> Result<String> {
-    Ok(node.lock().unwrap().get_gas_price().to_hex())
+async fn get_gas_price(_: Params<'_>, node: &Arc<Mutex<Node>>) -> Result<String> {
+    Ok(node.lock().await.get_gas_price().to_hex())
 }
 
-fn get_block_by_number(params: Params, node: &Arc<Mutex<Node>>) -> Result<Option<eth::Block>> {
+async fn get_block_by_number(
+    params: Params<'_>,
+    node: &Arc<Mutex<Node>>,
+) -> Result<Option<eth::Block>> {
     let mut params = params.sequence();
     let block_number: BlockNumber = params.next()?;
     let full: bool = params.next()?;
 
-    let node = node.lock().unwrap();
-    let block = node.get_block_by_number(block_number)?;
-
-    let block = block.map(|b| convert_block(&node, &b, full)).transpose()?;
+    let node = node.lock().await;
+    let block = node.get_block_by_number(block_number).await?;
 
-    Ok(block)
+    if let Some(block) = block {
+        Some(convert_block(&node, &block, full).await).transpose()
+    } else {
+        Ok(None)
+    }
 }
 
-fn get_block_by_hash(params: Params, node: &Arc<Mutex<Node>>) -> Result<Option<eth::Block>> {
+async fn get_block_by_hash(
+    params: Params<'_>,
+    node: &Arc<Mutex<Node>>,
+) -> Result<Option<eth::Block>> {
     let mut params = params.sequence();
     let hash: H256 = params.next()?;
     let full: bool = params.next()?;
 
-    let node = node.lock().unwrap();
-    let block = node
-        .get_block_by_hash(Hash(hash.0))?
-        .map(|b| convert_block(&node, &b, full))
-        .transpose()?;
-
-    Ok(block)
+    let node = node.lock().await;
+    if let Some(block) = node.get_block_by_hash(Hash(hash.0)).await? {
+        Some(convert_block(&node, &block, full).await).transpose()
+    } else {
+        Ok(None)
+    }
 }
 
-fn convert_block(node: &MutexGuard<Node>, block: &Block, full: bool) -> Result<eth::Block> {
+async fn convert_block(
+    node: &MutexGuard<'_, Node>,
+    block: &Block,
+    full: bool,
+) -> Result<eth::Block> {
     if !full {
         Ok(block.into())
     } else {
-        let transactions = block
-            .transactions
-            .iter()
-            .map(|h| {
-                get_transaction_inner(*h, node)?
-                    .ok_or_else(|| anyhow!("missing transaction: {}", h))
-            })
-            .map(|t| Ok(HashOrTransaction::Transaction(t?)))
-            .collect::<Result<_>>()?;
+        let transactions = try_join_all(block.transactions.iter().map(|h| async move {
```

Comment:
Actually on second thought I don't think splitting the function into a sync and an async part is worth it - since transaction fetching will likely also be extended to be async in the future, and `get_block_by_hash` is going to be cached in the node so should add very little overhead even if we fetch it again for every transaction.

I don't think the complexity here is excessive; it's very similar to e.g. a `Promise.all()` construction in javascript. Do you think it could benefit from some comments? Most of the lines of code are error handling (`.ok_or_else()`) and output shaping (converting into the `HashOrTransaction`, collecting into an appropriate vec, putting into an `eth::Block` object).

Reply:
It definetly seems excessively complex to me, I guess that is a matter of opinion. Could it not be something like:

```
        let mut transactions : Vec<HashOrTransaction> = vec![];

        for tx_hash in &block.transactions {
            let tx = get_transaction_inner(*tx_hash, node).await?.
                ok_or_else(|| anyhow!("missing transaction: {}", tx_hash))?;
            
            transactions.push(HashOrTransaction::Transaction(tx));
        }
```

Easier to breakpoint if something goes wrong imo, you are not looking at 100 partially resolved promises inside a closure/macro. IIRC rust documentation suggests to use a loop over a map when there are side effects.

As an aside, in the current code block when I look at it I am wondering why the last map returns an Ok?

Reply:
Hm, the tradeoff is whether the futures are executed sequentially or in parallel. I suppose with any sort of parallel construct then whenever something goes wrong you will end up mid-execution for all of them; doing it sequentially would certainly make it easier to breakpoint. However I'm not convinced this makes the cost worth it. I would assume that if anything goes wrong it would probably go wrong with one specific transaction, which should be easy enough to pinpoint even with the parallel construction with a bit of logging and then go from there, no?

If you still disagree then I'm open to changing it to just sequential execution. I'd be curious if @JamesHinshelwood has any opinion on this too, since we seem to have opposite preferences here.

> As an aside, in the current code block when I look at it I am wondering why the last map returns an Ok?

IIRC it's to coerce the entire thing to return a `Result<_>` using `collect()`, which in turn lets us use `?` to return early for any errors.

Reply:
@theo-zil In what situations would `get_transaction_inner` return `Poll::Pending` (i.e. not return immediately when `.await`ed)? Presumably it would occur when someone calls `eth_getBlock*` on a block/transactions the (light?) node hasn't downloaded yet and the future will complete when the block/transactions are downloaded? If my understanding is correct, I'd agree that *concurrent* (this isn't parallel) execution makes sense, because you can have requests for multiple transactions in the block (maybe ~1000 txns per block in the worst case?) pending at the same time.

Reply:
You're right, I was imprecisely using them as synonyms.

> In what situations would `get_transaction_inner` return `Poll:Pending`

Yep your assessment is correct - this will only happen when the block isn't locally stored so the node has to download it from the network. In the context this loop is in, we call `node.get_block().await` just above, guaranteeing that it would've just downloaded it (barring some _really_ bad cache eviction), but as you mention what we aren't guaranteed to have are the transactions.

That being said - and I've started a thread on slack to maybe brainstorm it a bit in an easier place than github discussion - I think we could simply include the transactions with `Message::BlockResponse`, which would again all but guarantee that they would be available here. Then sequential execution would be entirely fine here - I suppose the only benefit granted by making it concurrent would be if tokio schedules the futures on different threads and we multi-thread the processing (which may not necessarily even be worth its own overhead here).

I'll think about this a bit more and probably change it to sequential, honestly.

Reply:
Changed this to be sequential as per suggested snippet above.

Reply:
You started a thread on slack - is it private?

Reply:
I'll just post here for now if that's ok since I don't see any threads.

I am making two points here by the way, yes there is a question of paralell vs sequential, but there is just also a question of code complexity.

Ignoring the async stuff for now, consider the code blocks of what I suggested and what is (was) there. The complexity is significantly higher in terms of: how many return paths this function has (each ? is an additional return path), also more rust concepts, async move, multiple maps, collect and so on. KISS principle and all that :)

I notice when I look again that it probably doesn't work as intended with a try_join_all, the code path is that it will hit here:

```
pub(super) async fn get_transaction_inner(
    hash: Hash,
    node: &MutexGuard<'_, Node>,
) -> Result<Option<eth::Transaction>> {
    let Some(signed_transaction) = node.get_transaction_by_hash(hash)? else { return Ok(None); };
```

The get_transaction_by_hash will just check the mempool effectively, and return Ok(None), so no network events are made.

More generally, I would also say that even if that were what was desired, it feels like the cleanest implementation would be something like (broken pseudocode):

```
let transactions = node.get_transactions_from_hashes(?block.transactions).await?;
let transactions = transactions.map(|tx| HashOrTransaction::Transaction(tx));
```

Where 

```
// None if not all Transactions can be found
fn get_transactions_from_hashes(tx_hashes: &vec<Hash>) -> Option<Vec<Transaction>> {
   // Make sure we have all TXs
   let txs = mempool.get_transactions(tx_hashes);
   ...
}
```

Mempool:

```
async fn get_tranactions(tx_hashes: &vec<Hash>) -> Vec<Option<Transaction>> {
  let missing_transactions = tx_hashes.filter(|tx| {...local check...});
  if !missing_transactions.empty() {
      let missing_transactions = networking.get_txs(missing_transactions).await?;
  }
  ...
}
```

So in this scenario, you put the meat of the async logic at the lowest level, and instead of potentially creating one request per missing TX, you make it so its a vector that you request from the network. WDYT?

Reply:
Yeah I see what you mean. You're right that we currently don't have a mechanism to fetch transactions from the network, but the idea is that this will be added soon so I was generally considering the design as if they were.

But I agree that having a batching function at a lower level is probably a cleaner way to handle it: this would almost certainly not be the only place where we might need to make a batch of network requests for multiple items.